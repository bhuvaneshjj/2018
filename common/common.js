'use strict';

/* This file is present two times in the project repository:
   - common/common.js in CommonJS format (understood by Node JS)
   - ui/src/common.js in ES6 format (as required by create-react-app scripts)

   When Node.JS supports MJS modules formats natively, this could be just a single file.
 */

const PORT_DEV_CLIENT = 3000;
const PORT_DEV_SERVER = 4000;
const PLATBA_TYPY = ['hotově', 'převodem', 'složenkou'];

const BROADCAST_STOPKY = 'broadcastStopky';
const BROADCAST_UCASTNIK = 'broadcastUcastnik';
const DELETE_VYKON = 'deleteVykon';
const FIND_ALL_ROCNIKY = 'findAllRocniky';
const FIND_ALL_STOPKY = 'findAllStopky';
const FIND_ALL_UCASTNICI = 'findAllUcastnici';
const SAVE_PLATBY = 'savePlatby';
const SAVE_PRIHLASKA = 'savePrihlaska';
const SAVE_STOPKY = 'saveStopky';
const SAVE_UBYTOVANI = 'saveUbytovani';
const SAVE_UCAST = 'saveUcast';
const SAVE_UDAJE = 'saveUdaje';
const SAVE_VYKON = 'saveVykon';
const SIGN_IN = 'signIn';
const SIGN_OUT = 'signOut';

const CODE_OK = 'ok';
const CODE_ALREADY_EXISTING = 'již existuje';
const CODE_DB_DISCONNECTED = 'nepřipojeno k databázi';
const CODE_DUPLICIT_START_CISLO = 'duplicitní startovní číslo';
const CODE_KATEGORIE_INVALID = 'chybná kategorie';
const CODE_MAX_LOGIN_ATTEMPTS = 'max login attempts reached';
const CODE_MLADISTVY_UCASTNIK = 'účastník potřebuje souhlas zákonného zástupce';
const CODE_NONCE_MISMATCH = 'nesouhlas jednorázového přihlašovacího kódu';
const CODE_NONEXISTING = 'neexistuje';
const CODE_PASSWORD_INCORRECT = 'password incorrect';
const CODE_TOKEN_INVALID = 'authentication token invalid';
const CODE_UNFULFILLED_REQUEST = 'unfulfilled request';
const CODE_UNPARSEABLE_MESSAGE = 'unparseable message';
const CODE_UNRECOGNIZED_ACTION = 'unrecognized action';

const deleteVykon = ({ id, rok }, token) => ({
  action: DELETE_VYKON,
  request: { id, rok },
  token
});

const findAllRocniky = token => ({
  action: FIND_ALL_ROCNIKY,
  request: undefined,
  token
});

const findAllStopky = token => ({
  action: FIND_ALL_STOPKY,
  request: undefined,
  token
});

const findAllUcastnici = token => ({
  action: FIND_ALL_UCASTNICI,
  request: undefined,
  token
});

const savePlatby = ({ id, rok, platby }, token) => ({
  action: SAVE_PLATBY,
  request: { id, rok, platby },
  token
});

const savePrihlaska = ({ id, rok, prihlaska }, token) => ({
  action: SAVE_PRIHLASKA,
  request: { id, rok, prihlaska },
  token
});

const saveStopky = (stopky, token) => ({
  action: SAVE_STOPKY,
  request: stopky,
  token
});

const saveUbytovani = ({ id, rok, ubytovani }, token) => ({
  action: SAVE_UBYTOVANI,
  request: { id, rok, ubytovani },
  token
});

const saveUcast = ({ id, rok, udaje, prihlaska, vykon, platby, ubytovani, poznamka }, token) => ({
  action: SAVE_UCAST,
  request: { id, rok, udaje, prihlaska, vykon, platby, ubytovani, poznamka },
  token
});

const saveUdaje = ({ id, rok, udaje }, token) => ({
  action: SAVE_UDAJE,
  request: { id, rok, udaje },
  token
});

const saveVykon = ({ id, rok, vykon }, token) => ({
  action: SAVE_VYKON,
  request: { id, rok, vykon },
  token
});

const signIn = (username, password, nonce) => ({
  action: SIGN_IN,
  request: { username, password, nonce }
});

const signOut = token => ({
  action: SIGN_OUT,
  token
});

const zkontrolujMladistvy = (kategorie, { datum, narozeni, mladistvyPotvrzen }) => {
  if (!kategorie.vek) {
    return { kategorie, code: CODE_OK };
  }

  const PLNOLETOST = 18;
  const plnoletost = new Date(
    Date.UTC(narozeni.rok + PLNOLETOST, (narozeni.mesic || 1) - 1, narozeni.den || 0)
  );
  if (plnoletost > datum && !mladistvyPotvrzen) {
    return { kategorie: null, code: CODE_MLADISTVY_UCASTNIK };
  }

  return { kategorie, code: CODE_OK };
};

const filtrujPodleVeku = (kategorie, { rok, datum, narozeni }) =>
  kategorie.filter(item => {
    if (item.vek.presne) {
      const horniHranice = new Date(
        Date.UTC(narozeni.rok + item.vek.max + 1, (narozeni.mesic || 1) - 1, narozeni.den || 0)
      );
      if (horniHranice < datum) {
        return true;
      }
      return false;
    }

    const vek = rok - narozeni.rok;
    return item.vek.min <= vek && vek <= item.vek.max;
  });

const findKategorie = (rocniky, { rok, typ, pohlavi, narozeni, mladistvyPotvrzen }) => {
  const rocnik = rocniky[rok];
  if (!rocnik) {
    return { kategorie: null, code: CODE_NONEXISTING, status: `Ročník pro ${rok} neexistuje.` };
  }

  const typKategorie = rocnik.kategorie[typ];
  if (!typKategorie) {
    return {
      kategorie: null,
      code: CODE_NONEXISTING,
      status: `Kategorie '${typ}' v roce ${rok} neexistuje.`
    };
  }

  if (!typKategorie['muž'] && !typKategorie['žena']) {
    // jediná kategorie
    return zkontrolujMladistvy(typKategorie, {
      datum: new Date(rocnik.datum),
      narozeni,
      mladistvyPotvrzen
    });
  }

  const spravnePohlavi = typKategorie[pohlavi];
  if (!spravnePohlavi) {
    return {
      kategorie: null,
      code: CODE_NONEXISTING,
      status: `Kategorie '${typ}' v roce ${rok} neexistuje pro pohlaví '${pohlavi}'.`
    };
  }

  if (spravnePohlavi.length === 1) {
    return zkontrolujMladistvy(spravnePohlavi[0], {
      datum: new Date(rocnik.datum),
      narozeni,
      mladistvyPotvrzen
    });
  }

  if (!narozeni.rok) {
    return {
      kategorie: null,
      code: CODE_NONEXISTING,
      status: 'Nevyplněné datum narození (je potřeba alespoň rok).'
    };
  }

  const vek = rok - narozeni.rok;
  const spravnyVek = filtrujPodleVeku(spravnePohlavi, {
    rok,
    datum: new Date(rocnik.datum),
    narozeni
  });
  if (spravnyVek.length === 2 && spravnyVek[0].vek.presne) {
    spravnyVek.pop();
  }

  if (spravnyVek.length === 1) {
    return zkontrolujMladistvy(spravnyVek[0], {
      datum: new Date(rocnik.datum),
      narozeni,
      mladistvyPotvrzen
    });
  }

  if (spravnyVek.length === 0) {
    if (vek <= spravnePohlavi[0].vek.max) {
      return zkontrolujMladistvy(spravnePohlavi[0], {
        datum: new Date(rocnik.datum),
        narozeni,
        mladistvyPotvrzen
      });
    }
    return { kategorie: spravnePohlavi[spravnePohlavi.length - 1], code: CODE_OK };
  }

  return {
    kategorie: null,
    code: CODE_NONEXISTING,
    status: `Pro '${pohlavi}' - '${typ}' v roce ${rok} a věk ${vek} let jsem našel více kategorií.`
  };
};

const ubytovaniPrihlasit = ({ den, ubytovani = {} }) => ({
  ...ubytovani,
  [den]: { ...ubytovani[den], prihlaseno: true }
});
const ubytovaniOdhlasit = ({ den, ubytovani = {} }) => {
  const { [den]: remove, ...rest } = ubytovani;
  return rest;
};
const ubytovaniPrespano = ({ den, ubytovani = {} }) => ({
  ...ubytovani,
  [den]: { ...ubytovani[den], prespano: true }
});
const ubytovaniNeprespano = ({ den, ubytovani = {} }) => ({
  ...ubytovani,
  [den]: { ...ubytovani[den], prespano: false }
});

module.exports = {
  PORT_DEV_CLIENT,
  PORT_DEV_SERVER,
  PLATBA_TYPY,
  BROADCAST_STOPKY,
  BROADCAST_UCASTNIK,
  DELETE_VYKON,
  FIND_ALL_ROCNIKY,
  FIND_ALL_STOPKY,
  FIND_ALL_UCASTNICI,
  SAVE_PLATBY,
  SAVE_PRIHLASKA,
  SAVE_STOPKY,
  SAVE_UBYTOVANI,
  SAVE_UCAST,
  SAVE_UDAJE,
  SAVE_VYKON,
  SIGN_IN,
  SIGN_OUT,
  CODE_OK,
  CODE_ALREADY_EXISTING,
  CODE_DB_DISCONNECTED,
  CODE_DUPLICIT_START_CISLO,
  CODE_KATEGORIE_INVALID,
  CODE_MAX_LOGIN_ATTEMPTS,
  CODE_MLADISTVY_UCASTNIK,
  CODE_NONCE_MISMATCH,
  CODE_NONEXISTING,
  CODE_PASSWORD_INCORRECT,
  CODE_TOKEN_INVALID,
  CODE_UNFULFILLED_REQUEST,
  CODE_UNPARSEABLE_MESSAGE,
  CODE_UNRECOGNIZED_ACTION,
  deleteVykon,
  findAllRocniky,
  findAllStopky,
  findAllUcastnici,
  findKategorie,
  savePlatby,
  savePrihlaska,
  saveStopky,
  saveUbytovani,
  saveUcast,
  saveUdaje,
  saveVykon,
  signIn,
  signOut,
  ubytovaniNeprespano,
  ubytovaniOdhlasit,
  ubytovaniPrespano,
  ubytovaniPrihlasit
};
