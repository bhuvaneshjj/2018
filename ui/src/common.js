/* This file is present two times in the project repository:
   - common/common.js in CommonJS format (understood by Node JS)
   - ui/src/common.js in ES6 format (as required by create-react-app scripts)

   When Node.JS supports MJS modules formats natively, this could be just a single file.
 */

import moment from 'moment';

export const PORT_DEV_CLIENT = 3000;
export const PORT_DEV_SERVER = 4000;
export const PLATBA_TYPY = ['hotově', 'převodem', 'složenkou'];

export const API_DELETE_VYKON = 'deleteVykon';
export const API_FIND_ALL_ROCNIKY = 'findAllRocniky';
export const API_FIND_ALL_STOPKY = 'findAllStopky';
export const API_FIND_ALL_UCASTNICI = 'findAllUcastnici';
export const API_MODIFY_STOPKY = 'modifyStopky';
export const API_MODIFY_UBYTOVANI = 'modifyUbytovani';
export const API_SAVE_PLATBY = 'savePlatby';
export const API_SAVE_PRIHLASKA = 'savePrihlaska';
export const API_SAVE_UBYTOVANI = 'saveUbytovani';
export const API_SAVE_UCAST = 'saveUcast';
export const API_SAVE_UDAJE = 'saveUdaje';
export const API_SAVE_VYKON = 'saveVykon';
export const API_SIGN_IN = 'signIn';
export const API_SIGN_OUT = 'signOut';
export const BROADCAST_STOPKY = 'broadcastStopky';
export const BROADCAST_UCASTNIK = 'broadcastUcastnik';

export const CODE_OK = 'ok';
export const CODE_ALREADY_EXISTING = 'již existuje';
export const CODE_DB_DISCONNECTED = 'nepřipojeno k databázi';
export const CODE_DUPLICIT_START_CISLO = 'duplicitní startovní číslo';
export const CODE_KATEGORIE_INVALID = 'chybná kategorie';
export const CODE_MAX_LOGIN_ATTEMPTS = 'max login attempts reached';
export const CODE_MLADISTVY_UCASTNIK = 'účastník potřebuje souhlas zákonného zástupce';
export const CODE_NONCE_MISMATCH = 'nesouhlas jednorázového přihlašovacího kódu';
export const CODE_NONEXISTING = 'neexistuje';
export const CODE_NOT_ALLOWED = 'nepovoleno';
export const CODE_PASSWORD_INCORRECT = 'password incorrect';
export const CODE_TOKEN_INVALID = 'authentication token invalid';
export const CODE_UNFULFILLED_REQUEST = 'unfulfilled request';
export const CODE_UNPARSEABLE_MESSAGE = 'unparseable message';
export const CODE_UNRECOGNIZED_ACTION = 'unrecognized action';

export const apiCall = ({ endpoint, request, token }) => ({ action: endpoint, request, token });

const zkontrolujMladistvy = (kategorie, { datum, narozeni, mladistvyPotvrzen }) => {
  if (!kategorie.vek) {
    return { kategorie, code: CODE_OK };
  }

  const PLNOLETOST = 18;
  const plnoletost = new Date(
    Date.UTC(narozeni.rok + PLNOLETOST, (narozeni.mesic || 1) - 1, narozeni.den || 0)
  );
  if (plnoletost > datum && !mladistvyPotvrzen) {
    return { kategorie: null, code: CODE_MLADISTVY_UCASTNIK };
  }

  return { kategorie, code: CODE_OK };
};

const filtrujPodleVeku = (kategorie, { rok, datum, narozeni }) =>
  kategorie.filter(item => {
    if (item.vek.presne) {
      const horniHranice = new Date(
        Date.UTC(narozeni.rok + item.vek.max + 1, (narozeni.mesic || 1) - 1, narozeni.den || 0)
      );
      if (horniHranice < datum) {
        return true;
      }
      return false;
    }

    const vek = rok - narozeni.rok;
    return item.vek.min <= vek && vek <= item.vek.max;
  });

export const findKategorie = (rocniky, { rok, typ, pohlavi, narozeni, mladistvyPotvrzen }) => {
  const rocnik = rocniky[rok];
  if (!rocnik) {
    return { kategorie: null, code: CODE_NONEXISTING, status: `Ročník pro ${rok} neexistuje.` };
  }

  const typKategorie = rocnik.kategorie[typ];
  if (!typKategorie) {
    return {
      kategorie: null,
      code: CODE_NONEXISTING,
      status: `Kategorie '${typ}' v roce ${rok} neexistuje.`
    };
  }

  if (!typKategorie['muž'] && !typKategorie['žena']) {
    // jediná kategorie
    return zkontrolujMladistvy(typKategorie, {
      datum: new Date(rocnik.datum),
      narozeni,
      mladistvyPotvrzen
    });
  }

  const spravnePohlavi = typKategorie[pohlavi];
  if (!spravnePohlavi) {
    return {
      kategorie: null,
      code: CODE_NONEXISTING,
      status: `Kategorie '${typ}' v roce ${rok} neexistuje pro pohlaví '${pohlavi}'.`
    };
  }

  if (spravnePohlavi.length === 1) {
    return zkontrolujMladistvy(spravnePohlavi[0], {
      datum: new Date(rocnik.datum),
      narozeni,
      mladistvyPotvrzen
    });
  }

  if (!narozeni.rok) {
    return {
      kategorie: null,
      code: CODE_NONEXISTING,
      status: 'Nevyplněné datum narození (je potřeba alespoň rok).'
    };
  }

  const vek = rok - narozeni.rok;
  const spravnyVek = filtrujPodleVeku(spravnePohlavi, {
    rok,
    datum: new Date(rocnik.datum),
    narozeni
  });
  if (spravnyVek.length === 2 && spravnyVek[0].vek.presne) {
    spravnyVek.pop();
  }

  if (spravnyVek.length === 1) {
    return zkontrolujMladistvy(spravnyVek[0], {
      datum: new Date(rocnik.datum),
      narozeni,
      mladistvyPotvrzen
    });
  }

  if (spravnyVek.length === 0) {
    if (vek <= spravnePohlavi[0].vek.max) {
      return zkontrolujMladistvy(spravnePohlavi[0], {
        datum: new Date(rocnik.datum),
        narozeni,
        mladistvyPotvrzen
      });
    }
    return { kategorie: spravnePohlavi[spravnePohlavi.length - 1], code: CODE_OK };
  }

  return {
    kategorie: null,
    code: CODE_NONEXISTING,
    status: `Pro '${pohlavi}' - '${typ}' v roce ${rok} a věk ${vek} let jsem našel více kategorií.`
  };
};

export const STOPKY_ADD_MEZICAS = 'STOPKY_ADD_MEZICAS';
export const STOPKY_CHANGE_TIME = 'STOPKY_CHANGE_TIME';
export const STOPKY_INSERT_MEZICAS = 'STOPKY_INSERT_MEZICAS';
export const STOPKY_REMOVE_MEZICAS = 'STOPKY_REMOVE_MEZICAS';
export const STOPKY_RESET = 'STOPKY_RESET';
export const STOPKY_START = 'STOPKY_START';
export const STOPKY_STOP = 'STOPKY_STOP';
export const casSortMethod = (a, b) =>
  moment.duration(a.cas).asMilliseconds() - moment.duration(b.cas).asMilliseconds();

export const UBYTOVANI_PRIHLASIT = 'přihlásit';
export const UBYTOVANI_ODHLASIT = 'odhlásit';
export const UBYTOVANI_PRESPANO = 'přespáno';
export const UBYTOVANI_NEPRESPANO = 'nepřespáno';
export const ubytovaniModifications = {
  [UBYTOVANI_PRIHLASIT]: ({ den, ubytovani = {} }) => ({
    ...ubytovani,
    [den]: { ...ubytovani[den], prihlaseno: true }
  }),
  [UBYTOVANI_ODHLASIT]: ({ den, ubytovani = {} }) => {
    const { [den]: remove, ...rest } = ubytovani;
    return rest;
  },
  [UBYTOVANI_PRESPANO]: ({ den, ubytovani = {} }) => ({
    ...ubytovani,
    [den]: { ...ubytovani[den], prespano: true }
  }),
  [UBYTOVANI_NEPRESPANO]: ({ den, ubytovani = {} }) => ({
    ...ubytovani,
    [den]: { ...ubytovani[den], prespano: false }
  })
};
