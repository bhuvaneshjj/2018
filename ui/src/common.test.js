/* This file is present two times in the project repository:
   - common/common.test.js in CommonJS format (understood by Node JS)
   - ui/src/common.test.js in ES6 format (as required by create-react-app scripts)

   Convert from CommonJS format by the following incantation:
   cp common/common.test.js ui/src/common.test.js
   ./node_modules/jscodeshift/bin/jscodeshift.sh -t node_modules/5to6-codemod/transforms/cjs.js ui/src/common.test.js
   diff -u common/common.test.js ui/src/common.test.js
   diff -u common/__snapshots__/common.test.js.snap ui/src/__snapshots__/common.test.js.snap
 */

import { findKategorie } from './common';

it('findKategorie() - ročník neexistuje', () => {
  const rocniky = {};
  const nalezeno = findKategorie(rocniky, {
    rok: 2018,
    typ: 'maraton',
    pohlavi: 'muž',
    narozeni: { rok: 1956 }
  });
  expect(nalezeno.kategorie).toBeNull();
  expect(nalezeno).toMatchSnapshot();
});

it('findKategorie() - typ kategorie neexistuje', () => {
  const rocniky = { 2018: { kategorie: { půlmaraton: {} } } };
  const nalezeno = findKategorie(rocniky, {
    rok: 2018,
    typ: 'maraton',
    pohlavi: 'muž',
    narozeni: { rok: 1956 }
  });
  expect(nalezeno.kategorie).toBeNull();
  expect(nalezeno).toMatchSnapshot();
});

it('findKategorie() - jedna kategorie pro typ', () => {
  const rocniky = {
    2018: {
      datum: new Date(Date.UTC(2018, 5, 9, 0, 0, 0)),
      kategorie: { pěší: { id: '1234', typ: 'pěší' } }
    }
  };
  const nalezeno = findKategorie(rocniky, {
    rok: 2018,
    typ: 'pěší',
    pohlavi: 'muž',
    narozeni: { rok: 1956 }
  });
  expect(nalezeno.kategorie).toBeTruthy();
  expect(nalezeno).toMatchSnapshot();
});

it('findKategorie() - pohlaví neexistuje', () => {
  const rocniky = {
    2018: {
      datum: new Date(Date.UTC(2018, 5, 9, 0, 0, 0)),
      kategorie: { půlmaraton: { žena: [] } }
    }
  };
  const nalezeno = findKategorie(rocniky, {
    rok: 2018,
    typ: 'půlmaraton',
    pohlavi: 'muž',
    narozeni: { rok: 1956 }
  });
  expect(nalezeno.kategorie).toBeNull();
  expect(nalezeno).toMatchSnapshot();
});

it('findKategorie() - jedna kategorie pro pohlaví', () => {
  const rocniky = {
    2018: {
      datum: new Date(Date.UTC(2018, 5, 9, 0, 0, 0)),
      kategorie: {
        koloběžka: {
          muž: [{ id: '1234', typ: 'koloběžka', pohlavi: 'muž', vek: { min: 18, max: 150 } }],
          žena: [{ id: '3456', typ: 'koloběžka', pohlavi: 'žena', vek: { min: 18, max: 150 } }]
        }
      }
    }
  };
  const nalezeno = findKategorie(rocniky, {
    rok: 2018,
    typ: 'koloběžka',
    pohlavi: 'žena',
    narozeni: { rok: 1956 }
  });
  expect(nalezeno.kategorie).toBeTruthy();
  expect(nalezeno).toMatchSnapshot();
});

it('findKategorie() - nejmladší věková kategorie', () => {
  const rocniky = {
    2018: {
      datum: new Date(Date.UTC(2018, 5, 9, 0, 0, 0)),
      kategorie: {
        půlmaraton: {
          muž: [
            { id: '1234', vek: { min: 30, max: 39 } },
            { id: '2345', vek: { min: 40, max: 49 } },
            { id: '3456', vek: { min: 50, max: 59 } }
          ]
        }
      }
    }
  };
  const nalezeno = findKategorie(rocniky, {
    rok: 2018,
    typ: 'půlmaraton',
    pohlavi: 'muž',
    narozeni: { rok: 1999 }
  });
  expect(nalezeno.kategorie).toBeTruthy();
  expect(nalezeno).toMatchSnapshot();
});

it('findKategorie() - nejstarší věková kategorie', () => {
  const rocniky = {
    2018: {
      datum: new Date(Date.UTC(2018, 5, 9, 0, 0, 0)),
      kategorie: {
        půlmaraton: {
          muž: [
            { id: '1234', vek: { min: 40, max: 49 } },
            { id: '2345', vek: { min: 50, max: 59 } },
            { id: '3456', vek: { min: 59, max: 60 } }
          ]
        }
      }
    }
  };
  const nalezeno = findKategorie(rocniky, {
    rok: 2018,
    typ: 'půlmaraton',
    pohlavi: 'muž',
    narozeni: { rok: 1958 }
  });
  expect(nalezeno.kategorie).toBeTruthy();
  expect(nalezeno).toMatchSnapshot();
});

it('findKategorie() - správná věková kategorie', () => {
  const rocniky = {
    2018: {
      datum: new Date(Date.UTC(2018, 5, 9, 0, 0, 0)),
      kategorie: {
        půlmaraton: {
          muž: [
            { id: '1234', vek: { min: 40, max: 49 } },
            { id: '2345', vek: { min: 50, max: 59 } }
          ]
        }
      }
    }
  };
  const nalezeno = findKategorie(rocniky, {
    rok: 2018,
    typ: 'půlmaraton',
    pohlavi: 'muž',
    narozeni: { rok: 1968 }
  });
  expect(nalezeno.kategorie).toBeTruthy();
  expect(nalezeno).toMatchSnapshot();
});

it('findKategorie() - mladistvý pro maraton/věk', () => {
  const rocniky = {
    2018: {
      datum: new Date(Date.UTC(2018, 5, 9, 0, 0, 0)),
      kategorie: {
        maraton: {
          muž: [
            { id: '1234', pohlavi: 'muž', vek: { min: 18, max: 29 } },
            { id: '2345', pohlavi: 'muž', vek: { min: 30, max: 39 } },
            { id: '3456', pohlavi: 'muž', vek: { min: 40, max: 49 } },
            { id: '4567', pohlavi: 'muž', vek: { min: 50, max: 59 } }
          ]
        }
      }
    }
  };
  const nalezeno = findKategorie(rocniky, {
    rok: 2018,
    typ: 'maraton',
    pohlavi: 'muž',
    narozeni: { rok: 2001 }
  });
  expect(nalezeno.kategorie).toBeNull();
  expect(nalezeno).toMatchSnapshot();
});

it('findKategorie() - mladistvá pro koloběžku', () => {
  const rocniky = {
    2018: {
      datum: new Date(Date.UTC(2018, 5, 9, 0, 0, 0)),
      kategorie: {
        koloběžka: {
          muž: [{ id: '1234', pohlavi: 'muž', vek: { min: 18, max: 150 } }],
          žena: [{ id: '2345', pohlavi: 'žena', vek: { min: 18, max: 150 } }]
        }
      }
    }
  };
  const nalezeno = findKategorie(rocniky, {
    rok: 2018,
    typ: 'koloběžka',
    pohlavi: 'žena',
    narozeni: { rok: 2002 }
  });
  expect(nalezeno.kategorie).toBeNull();
  expect(nalezeno).toMatchSnapshot();
});

it('findKategorie() - junior a cyklo přesný věk - mladistvý', () => {
  const rocniky = {
    2018: {
      datum: new Date(Date.UTC(2018, 5, 9, 0, 0, 0)),
      kategorie: {
        cyklo: {
          muž: [
            { id: '1234', pohlavi: 'muž', vek: { min: 16, max: 17, presne: true } },
            { id: '2345', pohlavi: 'muž', vek: { min: 18, max: 29 } },
            { id: '3456', pohlavi: 'muž', vek: { min: 30, max: 39 } }
          ]
        }
      }
    }
  };
  const nalezeno = findKategorie(rocniky, {
    rok: 2018,
    typ: 'cyklo',
    pohlavi: 'muž',
    narozeni: { rok: 2001 }
  });
  expect(nalezeno.kategorie).toBeNull();
  expect(nalezeno).toMatchSnapshot();
});

it('findKategorie() - junior a cyklo přesný věk - mladistvý s potvrzením', () => {
  const rocniky = {
    2018: {
      datum: new Date(Date.UTC(2018, 5, 9, 0, 0, 0)),
      kategorie: {
        cyklo: {
          muž: [
            { id: '1234', pohlavi: 'muž', vek: { min: 16, max: 17, presne: true } },
            { id: '2345', pohlavi: 'muž', vek: { min: 18, max: 29 } },
            { id: '3456', pohlavi: 'muž', vek: { min: 30, max: 39 } }
          ]
        }
      }
    }
  };
  const nalezeno = findKategorie(rocniky, {
    rok: 2018,
    typ: 'cyklo',
    pohlavi: 'muž',
    narozeni: { rok: 2001 },
    mladistvyPotvrzen: true
  });
  expect(nalezeno.kategorie).toBeTruthy();
  expect(nalezeno).toMatchSnapshot();
});

it('findKategorie() - junior a cyklo přesný věk - mladistvý s potvrzením - těsně před hranicí', () => {
  const rocniky = {
    2018: {
      datum: new Date(Date.UTC(2018, 5, 9, 0, 0, 0)),
      kategorie: {
        cyklo: {
          muž: [
            { id: '1234', pohlavi: 'muž', vek: { min: 16, max: 17, presne: true } },
            { id: '2345', pohlavi: 'muž', vek: { min: 18, max: 29 } },
            { id: '3456', pohlavi: 'muž', vek: { min: 30, max: 39 } }
          ]
        }
      }
    }
  };
  const nalezeno = findKategorie(rocniky, {
    rok: 2018,
    typ: 'cyklo',
    pohlavi: 'muž',
    narozeni: { rok: 2000, mesic: 6, den: 8 },
    mladistvyPotvrzen: true
  });
  expect(nalezeno.kategorie).toBeTruthy();
  expect(nalezeno).toMatchSnapshot();
});

it('findKategorie() - junior a cyklo přesný věk - mladistvý - těsně na hranici', () => {
  const rocniky = {
    2018: {
      datum: new Date(Date.UTC(2018, 5, 9, 0, 0, 0)),
      kategorie: {
        cyklo: {
          muž: [
            { id: '1234', pohlavi: 'muž', vek: { min: 16, max: 17, presne: true } },
            { id: '2345', pohlavi: 'muž', vek: { min: 18, max: 29 } },
            { id: '3456', pohlavi: 'muž', vek: { min: 30, max: 39 } }
          ]
        }
      }
    }
  };
  const nalezeno = findKategorie(rocniky, {
    rok: 2018,
    typ: 'cyklo',
    pohlavi: 'muž',
    narozeni: { rok: 2000, mesic: 6, den: 9 },
    mladistvyPotvrzen: true
  });
  expect(nalezeno.kategorie).toBeTruthy();
  expect(nalezeno).toMatchSnapshot();
});

it('findKategorie() - junior a cyklo přesný věk - mladistvý - těsně za hranicí', () => {
  const rocniky = {
    2018: {
      datum: new Date(Date.UTC(2018, 5, 9, 0, 0, 0)),
      kategorie: {
        cyklo: {
          muž: [
            { id: '1234', pohlavi: 'muž', vek: { min: 16, max: 17, presne: true } },
            { id: '2345', pohlavi: 'muž', vek: { min: 18, max: 29 } },
            { id: '3456', pohlavi: 'muž', vek: { min: 30, max: 39 } }
          ]
        }
      }
    }
  };
  const nalezeno = findKategorie(rocniky, {
    rok: 2018,
    typ: 'cyklo',
    pohlavi: 'muž',
    narozeni: { rok: 2000, mesic: 6, den: 10 },
    mladistvyPotvrzen: true
  });
  expect(nalezeno.kategorie).toBeTruthy();
  expect(nalezeno).toMatchSnapshot();
});
