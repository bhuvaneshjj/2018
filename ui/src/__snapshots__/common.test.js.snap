// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`findKategorie() - jedna kategorie pro pohlaví 1`] = `
Object {
  "code": "ok",
  "kategorie": Object {
    "id": "3456",
    "pohlavi": "žena",
    "typ": "koloběžka",
    "vek": Object {
      "max": 150,
      "min": 18,
    },
  },
}
`;

exports[`findKategorie() - jedna kategorie pro typ 1`] = `
Object {
  "code": "ok",
  "kategorie": Object {
    "id": "1234",
    "typ": "pěší",
  },
}
`;

exports[`findKategorie() - junior a cyklo přesný věk - mladistvý - těsně na hranici 1`] = `
Object {
  "code": "ok",
  "kategorie": Object {
    "id": "2345",
    "pohlavi": "muž",
    "vek": Object {
      "max": 29,
      "min": 18,
    },
  },
}
`;

exports[`findKategorie() - junior a cyklo přesný věk - mladistvý - těsně za hranicí 1`] = `
Object {
  "code": "ok",
  "kategorie": Object {
    "id": "2345",
    "pohlavi": "muž",
    "vek": Object {
      "max": 29,
      "min": 18,
    },
  },
}
`;

exports[`findKategorie() - junior a cyklo přesný věk - mladistvý 1`] = `
Object {
  "code": "účastník potřebuje souhlas zákonného zástupce",
  "kategorie": null,
}
`;

exports[`findKategorie() - junior a cyklo přesný věk - mladistvý s potvrzením - těsně před hranicí 1`] = `
Object {
  "code": "ok",
  "kategorie": Object {
    "id": "1234",
    "pohlavi": "muž",
    "vek": Object {
      "max": 17,
      "min": 16,
      "presne": true,
    },
  },
}
`;

exports[`findKategorie() - junior a cyklo přesný věk - mladistvý s potvrzením 1`] = `
Object {
  "code": "ok",
  "kategorie": Object {
    "id": "1234",
    "pohlavi": "muž",
    "vek": Object {
      "max": 17,
      "min": 16,
      "presne": true,
    },
  },
}
`;

exports[`findKategorie() - mladistvá pro koloběžku 1`] = `
Object {
  "code": "účastník potřebuje souhlas zákonného zástupce",
  "kategorie": null,
}
`;

exports[`findKategorie() - mladistvý pro maraton/věk 1`] = `
Object {
  "code": "účastník potřebuje souhlas zákonného zástupce",
  "kategorie": null,
}
`;

exports[`findKategorie() - nejmladší věková kategorie 1`] = `
Object {
  "code": "ok",
  "kategorie": Object {
    "id": "1234",
    "vek": Object {
      "max": 39,
      "min": 30,
    },
  },
}
`;

exports[`findKategorie() - nejstarší věková kategorie 1`] = `
Object {
  "code": "ok",
  "kategorie": Object {
    "id": "3456",
    "vek": Object {
      "max": 60,
      "min": 59,
    },
  },
}
`;

exports[`findKategorie() - nevyplněné narození 1`] = `
Object {
  "code": "neexistuje",
  "kategorie": null,
  "status": "Nevyplněné datum narození (je potřeba alespoň rok).",
}
`;

exports[`findKategorie() - nevyplněné pohlaví 1`] = `
Object {
  "code": "neexistuje",
  "kategorie": null,
  "status": "Kategorie 'maraton' v roce 2018 neexistuje.",
}
`;

exports[`findKategorie() - pohlaví neexistuje 1`] = `
Object {
  "code": "neexistuje",
  "kategorie": null,
  "status": "Kategorie 'půlmaraton' v roce 2018 neexistuje pro pohlaví 'muž'.",
}
`;

exports[`findKategorie() - ročník neexistuje 1`] = `
Object {
  "code": "neexistuje",
  "kategorie": null,
  "status": "Ročník pro 2018 neexistuje.",
}
`;

exports[`findKategorie() - správná věková kategorie 1`] = `
Object {
  "code": "ok",
  "kategorie": Object {
    "id": "2345",
    "vek": Object {
      "max": 59,
      "min": 50,
    },
  },
}
`;

exports[`findKategorie() - typ kategorie neexistuje 1`] = `
Object {
  "code": "neexistuje",
  "kategorie": null,
  "status": "Kategorie 'maraton' v roce 2018 neexistuje.",
}
`;
